core/src/main/java/io/grpc/internal/DelayedStream.java

```diff
@@ -56,263 +54,228 @@
+    checkNotNull(reason, "reason");
+    boolean cancelNow;
+    ClientStreamListener listenerToClose = null;
+    synchronized (this) {
+      cancelNow = passThrough;
+      if (!cancelNow) {
+        error = reason;
+        listenerToClose = listener;
+        // TODO(ejona): Close InputStreams?
+        pendingCalls.clear();
       }
     }
-    if (listenerToBeCalled != null) {
-      Preconditions.checkState(streamToBeCancelled == null, "unexpected streamToBeCancelled");
-      listenerToBeCalled.closed(reason, new Metadata());
+    if (cancelNow) {
+      realStream.cancel(reason);
     }
-    if (streamToBeCancelled != null) {
-      streamToBeCancelled.cancel(reason);
+    if (listenerToClose != null) {
+      listenerToClose.closed(reason, new Metadata());
     }
   }
 
   @Override
   public void halfClose() {
-    if (startedRealStream == null) {
-      synchronized (this) {
-        if (startedRealStream == null) {
-          pendingHalfClose = true;
-          return;
-        }
+    delayOrExecute(new Runnable() {
+      @Override
+      public void run() {
+        realStream.halfClose();
       }
-    }
-    startedRealStream.halfClose();
+    });
   }
 
   @Override
-  public void request(int numMessages) {
-    if (startedRealStream == null) {
-      synchronized (this) {
-        if (startedRealStream == null) {
-          pendingFlowControlRequests += numMessages;
-          return;
+  public void request(final int numMessages) {
+    if (passThrough) {
+      realStream.request(numMessages);
+    } else {
+      delayOrExecute(new Runnable() {
+        @Override
+        public void run() {
+          realStream.request(numMessages);
         }
-      }
+      });
     }
-    startedRealStream.request(numMessages);
   }
 
   @Override
-  public void setCompressor(Compressor compressor) {
-    if (startedRealStream == null) {
-      synchronized (this) {
-        if (startedRealStream == null) {
-          this.compressor = compressor;
-          return;
-        }
+  public void setCompressor(final Compressor compressor) {
+    checkNotNull(compressor, "compressor");
+    delayOrExecute(new Runnable() {
+      @Override
+      public void run() {
+        realStream.setCompressor(compressor);
       }
-    }
-    startedRealStream.setCompressor(compressor);
+    });
   }
 
   @Override
   public void setDecompressor(Decompressor decompressor) {
-    if (startedRealStream == null) {
-      synchronized (this) {
-        if (startedRealStream == null) {
-          this.decompressor = decompressor;
-          return;
-        }
-      }
-    }
-    startedRealStream.setDecompressor(decompressor);
+    checkNotNull(decompressor, "decompressor");
+    // This method being called only makes sense after setStream() has been called (but not
+    // necessarily returned), but there is not necessarily a happens-before relationship. This
+    // synchronized block creates one.
+    synchronized (this) { }
```

Comment:
I purposefully didn't check realStream in the block to prevent being misleading. That's the same reason I didn't reference `passThrough` just to trigger the barrier. I don't want someone to think they understand the code when they don't.

I can't find an official, explicit specification that empty synchronized blocks behave as expected. However, I do believe it is not necessary to be explicit because they do have semantic behavior, like blocking and barriers, such that it is strictly not safe to remove them. One paper I saw only mentioned them in order to point out that you can't remove empty synchronized blocks. I also see multiple papers describing empty synchronized blocks working as expected, including in Java Concurrency in Practice page 272.